src/pool.rs

```diff
@@ -0,0 +1,35 @@
+use std::{sync, thread};
+
+pub(crate) type Task = dyn FnOnce() + Send;
+pub(crate) type BoxedTask = Box<Task>;
+
+pub(crate) struct ScopedPool<I>(I);
```

Comment:
Since this type is only ever used as `ScopedPool(vec_of_tasks).run(num_threads)` I would suggest deleting this type and the `<I>` type parameters. A single function without generics would be sufficient and more straightforward: `fn run_all(tasks: Vec<BoxedTask>, num_threads: Option<usize>)`.

In general this is a stylistic question that people can and will disagree with, but in this specific case I think the extra abstraction obscures what's actually going on. For example, it's useful to know that advancing the iterator doesn't do any significant work on its own (otherwise holding a lock while advancing it may be problematic) and that's not true for arbitrary `IntoIterator`s, but it's true for `Vec`.

Reply:
Right, there is no real need for a structure so I changed to a `fn`. However I left the `IntoIterator` for another review round, as I do see a valid reason for it to be generic. See the newly-added tests where we create dummy-tasks on-demand rather than pre-creating them to a Vec. True, we expect the operation to be cheap enough.

Reply:
If it really makes a difference, I could add a `pub(crate) fn` version which accepts `Vec` and delegates to the generic version which becomes private.

Reply:
I'm not sold on those tests (and thus, on keeping a more complicated interface for their sake). Recording thread IDs of everyone who tries to get a task from the iterator is a clever hack, but it *is* a hack, not a test of behavior that actually matters. The previous buggy version of this patch would have passed those tests, since it did call `next()` on the tasks iterator from every thread at least once! A hypothetical buggy version that pulled tasks from the iterator but didn't run them would also pass the test. Conversely, the tests nail down current implementation details and would unnecessarily fail if they changed (e.g., if the main thread consumed the iterator itself and pushed all tasks to a multi-consumer channel that the workers pull from). It's just really hard to have good unit tests for thread pools, especially for non-functional requirements such as "actually utilize all the threads".